istioctl/pkg/auth/validator.go

```diff
@@ -0,0 +1,122 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package auth
+
+import (
+	"fmt"
+	"strings"
+
+	rbacproto "istio.io/api/rbac/v1alpha1"
+	"istio.io/istio/pilot/pkg/model"
+)
+
+type Validator struct {
+	PolicyFiles          []string
+	RoleKeyToServiceRole map[string]model.Config
+	serviceRoleBindings  []model.Config
+	Report               strings.Builder
+}
+
+const (
+	RoleNotFound  = "serviceRoleNotFound: %q used by ServiceRoleBinding %q at namespace %q.\n"
+	RoleNotUsed   = "serviceRoleNotUsed: ServiceRole %q at namespace %q.\n"
+	PolicyValid   = "Authorization policy is valid.\n"
+	PolicyMissing = "needs to have both ServiceRole and ServiceRoleBinding for validation"
+)
+
+// CheckAndReport checks for Istio authentication and authorization mis-usage.
+func (v *Validator) CheckAndReport() error {
+	err := v.getRoleAndBindingLists()
+	if err != nil {
+		return err
+	}
+	v.CheckAndReportRBAC()
+	if v.Report.String() == "" {
+		v.Report.WriteString(GetPolicyValidReport())
+	}
+	return nil
+}
+
+func (v *Validator) CheckAndReportRBAC() {
+	usedRoleNames := map[string]bool{}
+	// Check if ServiceRoleBinding is using an non existent ServiceRole.
+	for _, binding := range v.serviceRoleBindings {
+		bindingSpec := binding.Spec.(*rbacproto.ServiceRoleBinding)
+		namespace := binding.Namespace
+		roleName := bindingSpec.RoleRef.Name
+		if v.doesRoleExist(namespace, roleName) {
+			roleKey := getRoleKey(namespace, roleName)
```

Comment:
NIT: move this line (construct roleKey) to before "if" statement, and change doesRoleExist() to take roleKey as input?

Reply:
Good idea. Will change